src/kubernetes/client/config/in_cluster.rs

```diff
@@ -0,0 +1,75 @@
+//! Everything related to building in-cluster configuration.
+
+use super::Config;
+use crate::tls::TlsOptions;
+use http02::Uri;
+use snafu::{ResultExt, Snafu};
+
+impl Config {
+    /// Prepares a config suitable for use when running in k8s cluster.
+    pub fn in_cluster() -> Result<Self, Error> {
+        let host = std::env::var("KUBERNETES_SERVICE_HOST").context(NotInCluster {
+            missing: "KUBERNETES_SERVICE_HOST",
+        })?;
+        let port = std::env::var("KUBERNETES_SERVICE_PORT").context(NotInCluster {
+            missing: "KUBERNETES_SERVICE_PORT",
+        })?;
+
+        let base = Uri::builder()
+            .scheme("https")
+            .authority(join_host_port(host.as_str(), port.as_str()).as_str())
+            .path_and_query("/")
+            .build()
+            .context(InvalidUrl)?;
+
+        let token_file = "/var/run/secrets/kubernetes.io/serviceaccount/token";
+        let root_ca_file = "/var/run/secrets/kubernetes.io/serviceaccount/ca.crt";
+
+        let token = std::fs::read_to_string(token_file).context(Token)?;
+
+        let mut tls_options = TlsOptions::default();
+        tls_options.ca_file = Some(root_ca_file.into());
+
+        Ok(Self {
+            base,
+            token,
+            tls_options,
+        })
+    }
+}
+
+/// An error returned when building an in-cluster configuration.
+#[derive(Debug, Snafu)]
+pub enum Error {
+    /// The in-cluster configuration requested while executing not in a cluster
+    /// environment.
+    #[snafu(display("unable to load in-cluster configuration, KUBERNETES_SERVICE_HOST and KUBERNETES_SERVICE_PORT must be defined"))]
+    NotInCluster {
+        /// The underlying error.
+        source: std::env::VarError,
+
+        /// The field that's missing.
+        missing: &'static str,
+    },
+
+    /// The token file could not be read successfully.
+    #[snafu(display("unable to read the token file"))]
+    Token {
+        /// The underlying error.
+        source: std::io::Error,
+    },
+
+    /// The configuration resulted in an invalid URL.
+    #[snafu(display("unable to construct a proper API server URL"))]
+    InvalidUrl {
+        /// The underlying error.
+        source: http02::Error,
+    },
+}
+
+fn join_host_port(host: &str, port: &str) -> String {
```

Comment:
I feel like there should be a safer way to do this with `url::Url::parse`?

Reply:
This is may not be a proper way to do it, but this is code is taken directly from the k8s source. I should probably add a comment about this. Also, I checked, and we don't currently have anything in rust that would work in this case.

I'll add documentation to this fn, and add tests to ensure it's working as expected.

Reply:
I'd rather you just used the existing production quality code that exists in `rust-url`. I'd be really upset to see us having unit tests for... url parsing.

Reply:
To be honest, even seeing us take `host, port` instead of a URL or addr of some kind is vaguely upsetting.

Reply:
Well, I'll have to add test cases either way - it's not *just a url*, but a very specific k8s format. Again, I've seen things at `rust-url` that would seem like thay're fit for the job, but in the end none of the built-in function worked - mainly because the logic is different. So, adding tests here would be beneficial to illustrate the edge-cases as a proof that we can't simply use `url` crate. Or, maybe, they'll help prove the opposite :smile: But I checked to see if I can use anything that already exists in rust, and I'm pretty sure I found that simply nothing fits.

Reply:
> To be honest, even seeing us take `host, port` instead of a URL or addr of some kind is vaguely upsetting.

That's the API that k8s provides: it passes service endpoints (L4 layer thing) information via two env vars: a host and a port.
That's a bummer, and I'd also prefer URLs. But then they communicate the parameters at the L4 (IP + TCP/UDP port), not a hight level - L7 - info - i.e. there's even no schema.

In a broader sense, the in-cluster config is implemented by the spec, and it's a bit unconventional, but it makes sense from the k8s point of view.

Also, it's a shame rust doesn't have this exact function (maybe with a different signature, but wit the same logic) in any of the popular crates. This interface is what people would want to use for service discovery at the k8s cluster (on par with DNS), and it's very easy to miss the nuance with IPv6 addresses support. I did this mistake a while ago :smile:

Reply:
Hmmm ok I didn't know it was a specific K8s format. That looks like ipv4/ipv6 port addition to me. :( Sorry.

Reply:
> Hmmm ok I didn't know it was a specific K8s format. That looks like ipv4/ipv6 port addition to me. :( Sorry.

It is! I couldn't find this in Rust though :frowning: Can `url` crate do that correctly? I'd prefer to use the standard method if there is one. `Url::parse` is pretty much what we do down the line, but the mere proper concatenation of host/port that accounts for IPv6 is nowhere to be found.

Reply:
Ah, this poor fn gets way too much attention. :smile: I've added code comments to explain the logic, although I feel like it's super trivial and readable as-is.
The doc-comment for the fn justifies the existence of the function and that there's no library fn in the language use. That's non-trivial, so I'll keep it.

Reply:
Indeed, poor fn.